File: android/app/src/main/java/com/dungeonboss/game/CrawlModifiers.kt
package com.dungeonboss.game

/**
 * Per-crawl modifiers produced by ability cards and discard-to-boost decisions
 * made just before a party crawls. They affect only that one crawl, and are
 * keyed by room index within the dungeon's encounters. Mirrors
 * `webapp/lib/crawl_modifiers.rb`.
 */
class CrawlModifiers {
    private val plus = HashMap<Int, Int>()          // +damage (Reinforcements / boost)
    private val zeroed = HashMap<Int, Boolean>()    // damage forced to 0 (Sabotage)
    private val setTo = HashMap<Int, Int>()         // damage overridden (discard-to-boost)
    private val unreducibleAt = HashMap<Int, Boolean>() // damage cannot be reduced
    private var retreatAt: Int? = null              // party turns back at this index (Retreat)

    /** A deep copy, so a play can be snapshotted and restored on undo. */
    fun copy(): CrawlModifiers {
        val c = CrawlModifiers()
        c.plus.putAll(plus)
        c.zeroed.putAll(zeroed)
        c.setTo.putAll(setTo)
        c.unreducibleAt.putAll(unreducibleAt)
        c.retreatAt = retreatAt
        return c
    }

    fun addDamage(index: Int, amount: Int = 2) {
        plus[index] = (plus[index] ?: 0) + amount
    }

    fun zero(index: Int) { zeroed[index] = true }
    fun setDamage(index: Int, value: Int) { setTo[index] = value }
    fun unreducibleMark(index: Int) { unreducibleAt[index] = true }
    fun retreat(index: Int) { retreatAt = index }

    fun retreating(): Boolean = retreatAt != null
    fun retreatIndex(): Int? = retreatAt
    fun retreatsAt(index: Int): Boolean = retreatAt?.let { index >= it } ?: false

    fun isZero(index: Int): Boolean = zeroed[index] ?: false
    fun isSet(index: Int): Boolean = setTo.containsKey(index)
    fun setValue(index: Int): Int = setTo.getValue(index)
    fun reducible(index: Int): Boolean = !(unreducibleAt[index] ?: false)
    fun bonus(index: Int): Int = plus[index] ?: 0

    /** Whether a per-room boost/override has already been applied (once per room). */
    fun boosted(index: Int): Boolean =
        setTo.containsKey(index) || (unreducibleAt[index] ?: false) ||
            (zeroed[index] ?: false) || (plus[index] ?: 0) > 0
}

File: webapp/lib/crawl_modifiers.rb
# frozen_string_literal: true

# Per-crawl modifiers produced by ability cards and discard-to-boost decisions
# made just before a party crawls. They affect only that one crawl. Keyed by
# room index within the dungeon's encounters.
class CrawlModifiers
  def initialize
    @plus = Hash.new(0)   # +damage (Reinforcements)
    @zero = {}            # damage forced to 0 (Sabotage)
    @set = {}             # damage overridden to a fixed value (discard-to-boost)
    @unreducible = {}     # damage cannot be reduced (Expose Weakness / Troll)
    @retreat_at = nil     # the party turns back at this encounter index (Retreat)
  end

  def add_damage(index, amount = 2) = @plus[index] += amount
  def zero!(index) = @zero[index] = true
  def set_damage!(index, value) = @set[index] = Integer(value)
  def boost6!(index) = set_damage!(index, 6) # convenience for the common case
  def unreducible!(index) = @unreducible[index] = true
  def retreat!(index) = @retreat_at = Integer(index) # crawl stops before this encounter

  def retreating? = !@retreat_at.nil?
  def retreat_index = @retreat_at
  def retreats_at?(index) = @retreat_at && index >= @retreat_at
  def zero?(index) = @zero.fetch(index, false)
  def set?(index) = @set.key?(index)
  def set_value(index) = @set[index]
  def reducible?(index) = !@unreducible.fetch(index, false)
  def bonus(index) = @plus[index]

  # Whether a per-room boost/override has already been applied (once per room).
  def boosted?(index)
    @set.key?(index) || @unreducible.fetch(index, false) ||
      @zero.fetch(index, false) || @plus[index].positive?
  end
end
